### DISCUSSION TITLE
Install OpenMPI during GitHub Action

### DISCUSSION THREAD

**Discussion by bissim** [community/community#26219] [Actions]
<p>Hello,</p>
<p>since I made more commits to fix my workflow than for my entire project, I’d appreciate some help.</p>
<p>I need to install OpenMPI before compiling my C project; here’s my workflow file:</p>
<pre><code>name: C CI

on: [push]

jobs:
  openmpi:
    runs-on: ubuntu-latest
    env:
      TOKEN: ${{ secrets.TOKEN }}
    steps:
    - name: download-openmpi
      run: wget https://download.open-mpi.org/release/open-mpi/v4.0/openmpi-4.0.2.tar.gz
    - name: extract-openmpi
      run: tar -xvf ./openmpi-4.0.2.tar.gz
    - name: configure-openmpi
      run: ./openmpi-4.0.2/configure --prefix="/home/${USER}/.openmpi"
    - name: install-openmpi
      run: |
        make -j
        sudo make install
  build:
    runs-on: ubuntu-latest
    needs: [openmpi]
    steps:
    - uses: actions/checkout@master
      with:
        fetch_depth: 1
    - name: set-openmpi-path
      run: |
        export PATH=${PATH}:/home/${USER}/.openmpi/bin/
        export LD_LIBRARY_PATH=${LD_LIBRARY_PATH}:/home/${USER}/.openmpi/lib/
        echo ${PATH}
        echo ${LD_LIBRARY_PATH}
    - name: make
      env:
        TOKEN: ${{ secrets.TOKEN }}
      run: make
</code></pre>
<p>Looks like all environment variables have been properly set and OpenMPI gets compiled and installed; however, in <em>build</em> job, I get the error about the <span class="hashtag">#include</span> “mpi.h” directive.</p>
<p>What am I missing?</p>

**Comment by Yanjingzhu** [comment, marked as answer]
<p>According to your workflow, you install OpenMPI and compile your C project in two separate jobs. <strong>Needs</strong> only deside the execution order of two jobs.  In Github Actions, different jobs will run on different runners, and the different runners are in different VMs. So you will get error when running job <strong>build</strong> , cause OpenMPI is not installed on that VM.</p>
<p>I would recommend you to install OpenMPI and compile your C project <strong>in one job</strong>.   </p>

**Comment by bissim** [comment]
<p>I guessed that was the case, despite hoping it didn’t work like that.</p>
<p>Thank you anyway, I merged the two jobs and make managed to find the OpenMPI C header.</p>
<p>Anyway, is there a way to organize workflows into phases sharing the same environment?</p>

**Comment by Yanjingzhu** [comment]
<p><em>&gt;&gt;Anyway, is there a way to organize workflows into phases sharing the same environment?</em></p>
<p>Do you want to run different jobs in the same environment? You could specify container for a job, then all steps without specify a container already will be executed in the container. You could create a docker image for your environment . Then use the same image in multiple jobs containers.  </p>
<p><a href="https://help.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobsjob_idcontainer" rel="nofollow noopener">https://help.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobsjob_idcontainer</a> </p>

**Comment by bissim** [comment]
<p>So, if I add</p>
<pre><code>container: ubuntu:latest
</code></pre>
<p>to openmpi job, the build one will run on the same environment as openmpi one?</p>

**Comment by airtower-luna** [comment]
<p>Unfortunately not, but if you run the job with a container image that includes the OpenMPI installation you need, it’ll already be available inside the container without having to install it first.</p>

**Comment by bissim** [comment]
<p>I see.</p>
<p>Thank you all for your answers!</p>
